agent/src/beerocks/slave/tasks/task_pool.cpp

```diff
@@ -0,0 +1,47 @@
+/* SPDX-License-Identifier: BSD-2-Clause-Patent
+ *
+ * SPDX-FileCopyrightText: 2016-2020 the prplMesh contributors (see AUTHORS.md)
+ *
+ * This code is subject to the terms of the BSD+Patent license.
+ * See LICENSE file for more details.
+ */
+
+#include "task_pool.h"
+
+using namespace beerocks;
+
+void TaskPool::add_task(std::shared_ptr<Task> new_task)
+{
+    m_task_pool[new_task->get_task_type()] = new_task;
```

Comment:
fixed.

Reply:
`add_task` should ask for a reference to `Task` instead of a pointer.
If you can't handle nullptr, **do not allow it in your interface**.

Reply:
> `add_task` should ask for a reference to `Task` instead of a pointer.
> If you can't handle nullptr, **do not allow it in your interface**.

Although I agree with the principle, the problem is that we need to express two things at the same time:
- if it is optional (i.e., can be nullptr);
- lifetime/ownership of the object.

Here, the shared_ptr is used to indicate co-ownership between the task pool and the caller - both of them can simultaneously keep the task alive. If we pass it as reference, this co-ownership is not possible.

I don't know if it is really *needed*, but it's definitely convenient to have the option.

Is there any other way to express this shared ownership? Or a pointer that is not allowed to be nullptr? Should we define a shared_ref type that is a shared_ptr but can't be nullptr?